scipy/optimize/direct.c

```diff
@@ -0,0 +1,86 @@
+#include "Python.h"
+#include "numpy/arrayobject.h"
+#include "direct.h"
+#define MY_ALLOC(p, t, n, ret_code) p = (t *) malloc(sizeof(t) * (n)); \
+                          if (!(p)) { *ret_code = DIRECT_OUT_OF_MEMORY; }
+#define MY_FREE(p) if (p) free(p)
```

Comment:
`MY_ALLOC` and `MY_FREE` looks like a bit of an antipattern. The `if (p)` part isn't needed for all `free`'s. For the rest, better to not do this and just write standard C constructs.

Reply:
As far as I can understand, you meant to replace `#define MY_FREE(p) if (p) free(p)` with `#define MY_FREE(p) free(p)` i.e., removing the `if (p)` check. Are you suggesting this because `if (p)` helps in hiding memory leaks which otherwise would have been detected because `free(p)` throws an error when `p = NULL`?

Reply:
No I meant "remove `MY_FREE`". The construct is correct, but the macro saving one line isn't ideal and calling anything `MY_*` is not really a good idea style-wise. I think it's preferable to just write out the C code as:
```C
if xxx:
    free(xxx)
```

Reply:
Ah yes in that sense you are right. I will remove `MY_FREE` altogether. I will do the same for `MY_ALLOC` as well. Thanks for explaining. :-)